### ISSUE TITLE
Macroexpansion with zero inputs fails inside quotations outside of definitions

### ISSUE THREAD

**Issue by timor** [factor/factor#2153]
I encountered this strange behavior, maybe someone can double-check that and tell me whether that is expected behavior?  Given the following MWE:

```
USING: kernel macros prettyprint ;

IN: foo


MACRO: foomacro ( x -- quot )
    [ . ] curry ;


: footest ( x -- x )
    42 foomacro ;

"footest:" print
1 footest drop

"manually:" print
1 42 [ foomacro ] call drop

MACRO: barmacro ( -- quot )
    [ 42 . ] ;

: bartest ( x -- x )
    barmacro ;

: bartest2 ( x -- x )
    [ barmacro ] call ;

"bartest:" print
1 bartest drop

"bartest2:" print
1 bartest2 drop

"manually:" print
1 [ barmacro ] call drop
```

All of the calls should print `42`, but the last one gives an error. Runnig this, I get the following output.

```
footest:
42
manually:
42
bartest:
42
bartest2:
42
manually:
Generic word nth-unsafe does not define a method for the hashtable class.
Dispatching on object: H{ }
```

The failing case seems to be:
1. The macro takes no inputs
2. The macro is used in a quotation outside of a definition

Version info: `Factor 0.99 x86.64 (1894, heads/master-7999e72aec, Jan  1 197...`

**Comment by mrjbq7** [comment]
Maybe an issue with the non-optimizing compiler? 

How are you running those calls at the bottom? From a listener? Or as a side effect to USE: foo?

**Comment by timor** [comment]
I ran it both from the listener, and with `fuel-run-file` from emacs.  Note that if you run `frontend` on the quotation, the output looks valid.

**Comment by mrjbq7** [comment]
Fixed in 0dc71ab053a67e1d15697ef28ac5018d52fbf1e1.
